UI/auth-evercast.cpp

```diff
@@ -0,0 +1,354 @@
+        };
+
+        return j;
+
+}
+
+nlohmann::json EvercastAuth::createStreamKeyQuery() {
+
+        nlohmann::json j;
+        j["query"] = "mutation getStreamKeyMutation(\n  $input: GetStreamKeyInput!\n) {\n  getStreamKey(input: $input) {\n    uuid\n  }\n}\n";
+        j["variables"] = {
+                {"input", {}}
+        };
+        return j;
+
+}
+
+nlohmann::json EvercastAuth::createRoomsQuery() {
+
+        nlohmann::json j;
+        j["query"] = "query homeQuery {\n"
+		"\tcurrentProfile {\n"
+		"\t\t...Home_currentProfile\n"
+		"\t}\n"
+		"\tcanCreateRoom\n"
+		"}\n"
+		"\n"
+		"fragment Home_currentProfile on Profile {\n"
+		"\n"
+		"\trooms: liveroomsByCreatorId(first: 100, orderBy: [CREATED_AT_DESC]) {\n"
+		"\t\tnodes {\n"
+		"\t\t\tid\n"
+		"\t\t\thash\n"
+		"\t\t\tcreatedAt\n"
+		"\t\t\tdeletedAt\n"
+		"\t\t\t...RoomCard_room\n"
+		"\t\t}\n"
+		"\t}\n"
+		"\tinvites: invitesByProfileId(first: 100, orderBy: [CREATED_AT_DESC]) {\n"
+		"\t\tnodes {\n"
+		"\t\t\tliveroomByRoomId {\n"
+		"\t\t\t\tid\n"
+		"\t\t\t\tcreatedAt\n"
+		"\t\t\t\tdeletedAt\n"
+		"\t\t\t\t...RoomCard_room\n"
+		"\t\t\t\tprofileByCreatorId {\n"
+		"\t\t\t\t\tdisplayName\n"
+		"\t\t\t\t}\n"
+		"\t\t\t}\n"
+		"\t\t}\n"
+		"\t}\n"
+		"\trecentRooms: joinedRoomsByProfileId(first: 100, orderBy: [JOINED_AT_DESC]) {\n"
+		"\t\tnodes {\n"
+		"\t\t\tjoinedAt\n"
+		"\t\t\tliveroomByRoomId {\n"
+		"\t\t\t\tid\n"
+		"\t\t\t\tcreatedAt\n"
+		"\t\t\t\tdeletedAt\n"
+		"\t\t\t\t...RoomCard_room\n"
+		"\t\t\t}\n"
+		"\t\t}\n"
+		"\t}\n"
+		"}\n"
+		"\n"
+		"fragment RoomCard_room on Liveroom {\n"
+		"\tid\n"
+		"\thash\n"
+		"\tdisplayName\n"
+		"\tcreatorId\n"
+		"\tcreatedAt\n"
+		"\tsessionCount\n"
+		"\tnoteCount\n"
+		"}";
+
+        return j;
+
+}
+
+EvercastAuth::Token EvercastAuth::obtainToken(const Credentials& credentials) {
+
+	httplib::Client client("https://v2.evercast.us");
+
+        auto query = createLoginQuery(credentials);
+
+        auto res = client.Post("/api/graphql", query.dump(), "application/json");
+
+        if (res) {
+                return getTokenInfoFromCookies(res->headers);
+        }
+
+	return {};
+
+}
+
+std::string EvercastAuth::obtainStreamKey(const Token& token) {
+
+        httplib::Client client("https://v2.evercast.us");
+
+        auto query = createStreamKeyQuery();
+
+        httplib::Headers headers({
+                                         {"X-Double-Submit", token.nonce},
+                                         {"cookie", "__Host-nonce=" + token.nonce + "; __Host-jwt=" + token.token}
+                                 });
+
+        auto res = client.Post("/api/graphql", headers, query.dump(), "application/json");
+
+        if (res) {
+                auto j = nlohmann::json::parse(res->body);
+                return j["data"]["getStreamKey"]["uuid"];
+        }
+
+	return "";
+
+}
+
+EvercastAuth::Rooms EvercastAuth::obtainRooms(const Token& token) {
+
+        httplib::Client client("https://v2.evercast.us");
```

Comment:
There's a config value we're using in UI/Update/update.cpp.  We should merge the base URL so that it's always pulled from config, with a default value of v2.evercast.us.

We may also be moving away from v2 - can you reach out to Castro (Guilherme Castro) to confirm whether the API will still be available at v2 a month from now?

We should also have a UI option to change this target during sign-on; it should be hidden by default.

Reply:
Silly me, of course this should come from the URL defined in settings - you will need to extract the host component of course but otherwise it should be serviceable.  It should also default to the current value - wss://v2.evercast.us/websockets, which I believe it currently doesn't.

Reply:
Please check the PR to this branch - https://github.com/evercast/evercast-broadcaster-software/pull/45v